Disallow manual combo colouring when not on beatmap skin

frenzibyte opened the pull request:
Closes #8282 

Disallows manual combo colouring (aka "color haxing") when not enabling beatmap skin, this is a behaviour on osu!stable and should be a thing here to avoid selecting combo colours of a different skin to hit objects that are not meant to be coloured in that way.

## Breaking changes
### `IHasComboInformation.ComboIndex` and `IHasComboInformation.ComboIndexBindable` have changed types from `int` to `ComboIndex (struct)`
Now `ComboIndex` provides two indices within its structure, automated index which increments by `1` per new combo, and manual index which increments by `1 + ComboOffset` per new combo, to allow switching between manual combo colouring and automated combo colouring.

Commit:
Separate combo skipping to another index within ComboIndex

Commit:
Disallow manual combo colouring when not on beatmap skin

Commit:
Add manual combo colouring test cases

Commit:
Fix failing test

bdach commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,71 @@
+        /// This is separated from <see cref="Manual"/> to disallow
+        /// manual combo colouring when not on the right skin to apply.
+        /// </remarks>
+        public readonly int Automated;
```
Not too sure about these names. Would rename to `Ordinal` (or `WithoutOffset`?)...

iiSaLMaN replied:
+1 for `Ordinal`, my english dictionary is still small :P

bdach commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,71 @@
+        /// <summary>
+        /// The manual index value of the combo, this accounts for <see cref="IHasCombo.ComboOffset"/>.
+        /// </summary>
+        public readonly int Manual;
```
...and this one to `WithOffset`.

iiSaLMaN replied:
`Manual` goes well with `Ordinal` as this index can be set to literally anything, but since it should take value from `IHasCombo.ComboOffset`, `WithOffset` sounds fair.

bdach commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,71 @@
+        public readonly int Automated;
+
+        /// <summary>
+        /// The manual index value of the combo, this accounts for <see cref="IHasCombo.ComboOffset"/>.
```
```suggestion
        /// The manual index value of the combo. Accounts for <see cref="IHasCombo.ComboOffset"/>.
```

bdach commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,71 @@
+    public readonly struct ComboIndex : IEquatable<ComboIndex>
+    {
+        /// <summary>
+        /// The automated index value of the combo in relation to the beatmap.
```
```suggestion
        /// The automated index value of the combo in relation to the beatmap. Ignores <see cref="IHasCombo.ComboOffset"/>.
```

bdach commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,71 @@
+        public static ComboIndex Add(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated + b.Automated, a.Manual + b.Manual);
+        public static ComboIndex Subtract(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated - b.Automated, a.Manual - b.Manual);
+        public static ComboIndex Multiply(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated * b.Automated, a.Manual * b.Manual);
+        public static ComboIndex Divide(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated / b.Automated, a.Manual / b.Manual);
```
I don't think these make a lot of sense (why would adding two combo indices add them element-wise?). Do they need to exist? I don't think there is *any* sane way to define arithmetic operators for this type.

iiSaLMaN replied:
I've added these such that the structure can almost act as a normal integer, such that custom implementations that need to interact / modify the combo index (say combo tests) can do so easily, for example:
```csharp
public void ApplyToHitCircle(HitCircle circle)
{
    Scheduler.AddDelayed(circle.ComboIndex += 2, 250);
}
```
since combo index is initialized as `(0, 0)`, adding 2 would do for both indices, and the combo index will become `(2, 2)` which will pick the third combo colour no matter the conditions. (as if you're adding 2 to an `int`)

bdach replied:
Yes, I get the idea. Thing is, with the current implementation, you also allow things like
```csharp
new ComboIndex(2, 6) + new ComboIndex(1, 5)
```
which will yield `new ComboIndex(3, 11)` but what I'm trying to get across is that *it doesn't really make much sense to do that semantically*. Those operations are ill-defined and *should not be possible to do in an implicit manner*. Even worse are the division operators. I don't think being able to do division on two instances of `ComboIndex` is going to ever be defensible.

Point is, (a) those are at this stage unused, and I generally discourage from trying to future-proof as it just produces dead code that might have unknown bugs, and (b) the API for manipulation of combo indices here is not really sane in many cases. Operator overloading in general I also think should be used *very* sparingly. At most I'd allow
```csharp
public struct ComboIndex
{
    public void Add(int amount);
    public void Subtract(int amount);
}
```
but even *then* strange issues can arise (what if you try to do `new ComboIndex(4, 0).Subtract(2)`? `ComboIndex(2, -2)` isn't really sane either, is it?)

iiSaLMaN replied:
Went ahead with static `Add` and `Subtract` methods only with filled-up XMLDocs to avoid misusing it, I suppose that's the least sane way.

bdach commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,71 @@
+        public static ComboIndex Divide(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated / b.Automated, a.Manual / b.Manual);
+
+        public static ComboIndex Increment(ComboIndex ci) => ci + 1;
+        public static ComboIndex Decrement(ComboIndex ci) => ci - 1;
```
These are supremely convoluted. To recap, you're relying *both* on the implicit conversion from `int` and the half-well-defined element-wise addition of combo indices to make this work. (And even then it's questionable if the increment operator should increment both members.)

iiSaLMaN replied:
see the other conversation above's reply, if you want to increment a specific index, you can just increment it directly `ComboIndex.Automated++`.

bdach reviewed (changes requested):
I think I understand the issue this aims to fix. I am unsure if this is really what stable does, or if it's a behaviour we even want to keep, as it over-complicates matters significantly. I can't see what stable does easily without reverse-engineering, so I'll defer on that matter to the team.

Some gripes with the code, however.

frenzibyte committed:
Adjust XMLDoc

Co-Authored-By: Bartłomiej Dach <dach.bartlomiej@gmail.com>

Commit:
Revert test fix

huoyaoyuan commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,71 @@
+        public static ComboIndex Add(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated + b.Automated, a.Manual + b.Manual);
+        public static ComboIndex Subtract(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated - b.Automated, a.Manual - b.Manual);
+        public static ComboIndex Multiply(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated * b.Automated, a.Manual * b.Manual);
+        public static ComboIndex Divide(ComboIndex a, ComboIndex b) => new ComboIndex(a.Automated / b.Automated, a.Manual / b.Manual);
```
Strongly disagree against method alternatives of operators.
Disagree with meaningless operations.

Proposed:
```csharp
public struct ComboIndex
{
    public static ComboIndex operator+(ComboIndex left, int count);
    public static ComboIndex operator-(ComboIndex left, int count);
}
```

iiSaLMaN replied:
method alternatives for operators are actually inspection rules, also duplicate of what @bdach said.

Commit:
Automated -> Ordinal, Manual -> WithOffset

'WithOffset' used as it reads well on usage: int index = comboIndex.WithOffset;

Commit:
Remove operator overloading and unnecessary arithmetic methods

Does not make sense semantically and mostly dead code.

Commit:
Add XMLDocs

Commit:
Merge branch 'master' into fix-beatmap-current-skin-combo-colours

huoyaoyuan commented:
Note: some CLS compliant rules or inspection rules are inherited from 2005, at which time VB had less features. Nowadays, just considering modern C# consumption is enough.

peppy commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,92 @@
+        /// This is separated from <see cref="WithOffset"/> to disallow
+        /// manual combo colouring when not on the right skin to apply.
+        /// </remarks>
+        public readonly int Ordinal;
```
so is this the combo's index in the beatmap? if so, you should also look at `IHasCombo.ComboOffset` as it may be similar to what is provided here.

as it stands, this documentation/class feels very hard to understand.

iiSaLMaN replied:
isn't `IHasCombo.ComboOffset` just the amount of combos to skip from current combo index in the processing? this index doesn't let that in its calculation: `Ordinal = lastObj.Ordinal + 1;` (which just increments by 1 per new combo, or as you said "the combo's index in the beatmap")

If osu!stable provides something much saner / straightforward than this, would be better to let me know I suppose.

peppy commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,92 @@
+        /// </summary>
+        /// <remarks>
+        /// This is separated from <see cref="WithOffset"/> to disallow
+        /// manual combo colouring when not on the right skin to apply.
```
"not on the right skin to apply"

i don't know what this means

Commit:
Adjust xmldocs

bdach reviewed (commented):
No major mechanical objections to the code itself, although again, I'm not 100% sure if this is the behaviour stable has or if we want to preserve it. (We probably do, but it's a little complicated to keep track of in the way done here; maybe this logic should be somehow encapsulated in legacy beatmap decoding?)

peppy commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,92 @@
+        public readonly int Ordinal;
+
+        /// <summary>
+        /// The manual index value of the combo. Accounts for <see cref="IHasCombo.ComboOffset"/>.
```
what does manual even mean here? i'd probably remove the word.

peppy commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,92 @@
+        /// <summary>
+        /// The manual index value of the combo. Accounts for <see cref="IHasCombo.ComboOffset"/>.
+        /// </summary>
+        public readonly int WithOffset;
```
`WithBeatmapOffset`

peppy commented on osu.Game/Rulesets/Objects/Types/ComboIndex.cs:
```diff
@@ -0,0 +1,92 @@
+namespace osu.Game.Rulesets.Objects.Types
+{
+    /// <summary>
+    /// Represents an <see cref="Ordinal"/> and <see cref="WithOffset"/> combo indices.
```
this isn't english nor does it describe the class. don't describe something using itself.

iiSaLMaN replied:
I've reworded it, please check if it's good now.

Commit:
Merge remote-tracking branch 'upstream/master' into fix-beatmap-current-skin-combo-colours

Commit:
Reword provided documentation

Commit:
WithOffset -> WithBeatmapOffset

Commit:
WithOffset -> WithBeatmapOffset (2)

frenzibyte committed:
Adjust inline comment

Co-Authored-By: Bartłomiej Dach <dach.bartlomiej@gmail.com>

iiSaLMaN commented:
> No major mechanical objections to the code itself, although again, I'm not 100% sure if this is the behaviour stable has or if we want to preserve it. (We probably do, but it's a little complicated to keep track of in the way done here; maybe this logic should be somehow encapsulated in legacy beatmap decoding?)

I think this is the best approach FWIW. encapsulating in legacy beatmap decoder isn't clear to me, can you elaborate further?

bdach commented:
The loose idea would be to check if the beatmap has a skin at parse time and then expose a `ComboIndex` or something at the legacy beatmap level if it does have a skin (so basically have it in `IHasCombo`, maybe?) and then use it only in skinning. I haven't looked in detail if that's viable, however.

iiSaLMaN commented:
> The loose idea would be to check if the beatmap has a skin at parse time and then expose a `ComboIndex` or something at the legacy beatmap level if it does have a skin (so basically have it in `IHasCombo`, maybe?) and then use it only in skinning. I haven't looked in detail if that's viable or if it'll look good in code, however.

so you're simply saying to move `ComboIndex.WithBeatmapOffset` to `IHasCombo` and set it on parse time, that almost sounds like what this PR does but slightly different and on parse time, which is incorrect because the combo index processing must be inside `BeatmapProcessor` for beatmap editing purposes, unless I'm mistaking?

frenzibyte commented:
I have another way that doesn't affect the `IHasComboInformation` interface like that, going to close this as it's gonna be a complete rewrite.
